### ISSUE TITLE
about the choice of p_th

### ISSUE THREAD

**Issue by Renp1ngs** [qinenergy/cotta#16]
Awesome work, but a little question: 

> Therefore, we use a threshold $p_{th}$ to filter the images, and do not apply augmentations on those with high confidence. More specifically, we design $p_{th} = conf^S-\delta$,  where $conf^S$ is the 5% quantile for the softmax predictions’ confidence on the source images from the source model $f_\theta$

In your code, 0.74 was the 5% quantile for cityscapes. So ,how is this value calculated?

Kind regards.

**Comment by qinenergy** [comment]
Hi, as mentioned in the supplementary, it is computed based on the source pre-trained (no adaptation) model's confidence on the source dataset. You feed in the source images, and get a bunch of confidence numbers. You sort them and get the 5% quantile number $s$. finally, $p_th$ is defined as $s-0.05$ in our work.

You will find that the number can differ quite a lot because of the datasets and the way the source network is trained. You can find the number for different datasets we pre-calculated in the config files in OPTIM:  AP. For example, in the above Cityscapes example, we use $p_th=0.74-0.05=0.69$.

**Comment by Renp1ngs** [comment]
Very clear answer. 
Best.

**Comment by Renp1ngs** [comment]
And a liitle question about the Stochastic Restoration in the code:

```python
        for nm, m  in model.named_modules():
            for npp, p in m.named_parameters():
                if npp in ['weight', 'bias'] and p.requires_grad:
                    mask = (torch.rand(p.shape)<0.01).float().cuda() 
                    with torch.no_grad():
                        p.data = anchor[f"{nm}.{npp}"] * mask + p * (1.-mask)
the file path is mmseg/apis/test.py.

So, can i wirten as :
```python
for npp, p in m.named_parameters():
      if npp in ['weight', 'bias'] and p.requires_grad:
         mask = (torch.rand(p.shape)<0.01).float().cuda() 
         with torch.no_grad():
                p.data = anchor[f"{nm}.{npp}"] * mask + p * (1.-mask)

**Comment by qinenergy** [comment]
Yes, you can. I think they are equivalent.

**Comment by Renp1ngs** [comment]
Thanks your reply.

**Comment by Renp1ngs** [comment]
Sorry, here I go again. 
When I extended your method, I found that the prediction used in calculating miou is derived from the teacher model without updated, not the updated student model. Have you tried to calculate the final miou with the updated student model?

**Comment by qinenergy** [comment]
> Sorry, here I go again. When I extended your method, I found that the prediction used in calculating miou is derived from the teacher model without updated, not the updated student model. Have you tried to calculate the final miou with the updated student model?

The teacher model can be seen as a slower-updated average of the student model. Therefore, if you use the same hyper-parameters, using the student model instead could have a lower performance in the long-term than using the teacher model. However, it is possible that you can overcome this by adjusting the learning rate or the forgetting rate.

**Comment by qinenergy** [comment]
Yes, this design choice is to keep the model change more smoothly in case of 1> sudden domain change 2> long term adaptation.
